### PR TITLE
Replace AWS SDK v2 with AWS SDK v3

### PR DESCRIPTION
**Pull request by jtoric**

### CODE REVIEW

File: server/shared/storage/providers/amazon.js [hunk 1]
```diff
@@ -1,13 +1,16 @@
 'use strict';
 
+const { CopyObjectCommand, DeleteObjectCommand, DeleteObjectsCommand, GetObjectCommand, HeadObjectCommand, ListObjectsV2Command, PutObjectCommand, S3Client } = require('@aws-sdk/client-s3');
```
Comment (droguljic) [inline server/shared/storage/providers/amazon.js]: Make this a multiline import,
```
const {
  CopyObjectCommand, 
  DeleteObjectCommand, 
  DeleteObjectsCommand, 
  GetObjectCommand, 
  HeadObjectCommand, 
  ListObjectsV2Command, 
  PutObjectCommand, 
  S3Client 
} = require('@aws-sdk/client-s3');
```

File: server/shared/storage/providers/amazon.js [hunk 2]
```diff
@@ -22,17 +25,16 @@ class Amazon {
     config = validateConfig(config, schema);
 
     const s3Config = {
-      signatureVersion: 'v4',
-      accessKeyId: config.key,
-      secretAccessKey: config.secret,
-      region: config.region,
-      apiVersion: '2006-03-01',
-      maxRetries: 3
```
Comment (droguljic) [inline server/shared/storage/providers/amazon.js:30]: What is with this property, can we set it for the compatibility reasons.

File: server/shared/storage/providers/amazon.js [hunk 3]
```diff
@@ -22,17 +25,16 @@ class Amazon {
     config = validateConfig(config, schema);
 
     const s3Config = {
-      signatureVersion: 'v4',
```
Comment (droguljic) [inline server/shared/storage/providers/amazon.js:25]: To which version is the signature set in the new client? We must ensure compatibility.

File: server/shared/storage/providers/amazon.js [hunk 4]
```diff
@@ -1,13 +1,16 @@
 'use strict';
 
+const { CopyObjectCommand, DeleteObjectCommand, DeleteObjectsCommand, GetObjectCommand, HeadObjectCommand, ListObjectsV2Command, PutObjectCommand, S3Client } = require('@aws-sdk/client-s3');
+const { createRequest } = require('@aws-sdk/s3-request-presigner');
+const { formatUrl } = require('@aws-sdk/util-format-url');
 const miss = require('mississippi');
 const path = require('path');
-const S3 = require('aws-sdk/clients/s3');
+const { Upload } = require('@aws-sdk/lib-storage');
```
Comment (droguljic) [inline server/shared/storage/providers/amazon.js:16]: Group with other AWS imports.

File: server/shared/storage/providers/amazon.js [hunk 5]
```diff
@@ -47,8 +49,8 @@ class Amazon {
   // API docs: http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#getObject-property
   getFile(key, options = {}) {
     const params = Object.assign(options, { Bucket: this.bucket, Key: key });
-    return this.client.getObject(params).promise()
-      .then(({ Body: data }) => data)
+    return this.client.send(new GetObjectCommand(params))
+      .then(({ Body: data }) => data.transformToString())
```
Comment (droguljic) [inline server/shared/storage/providers/amazon.js]: Before data was a buffer, and that is expected by consumers, so we should maintain compatibility and return buffer from the new implementation as well.
`Buffer.from(data.transformToByteArray())`

File: server/shared/storage/proxy/providers/cloudfront.js [hunk 6]
```diff
@@ -53,6 +59,13 @@ class CloudFront {
   getCookieNames() {
     return Object.values(storageCookies);
   }
+
+  _sign(policy) {
```
Comment (droguljic) [inline server/shared/storage/proxy/providers/cloudfront.js]: This should be moved from class to the end of file and renamed to `signPolicy`.

**Review by droguljic — CHANGES_REQUESTED** [review]
Also update references to AWS docs.

File: server/shared/storage/providers/amazon.js (hunk 1 above)
Comment (jtoric) [reply server/shared/storage/providers/amazon.js]: Fixed

File: server/shared/storage/providers/amazon.js (hunk 2 above)
Comment (jtoric) [reply server/shared/storage/providers/amazon.js:30]: Fixed

File: server/shared/storage/providers/amazon.js (hunk 3 above)
Comment (jtoric) [reply server/shared/storage/providers/amazon.js:25]: Fixed

File: server/shared/storage/providers/amazon.js (hunk 4 above)
Comment (jtoric) [reply server/shared/storage/providers/amazon.js:16]: Cannot do this because of eslint

File: server/shared/storage/providers/amazon.js (hunk 5 above)
Comment (jtoric) [reply server/shared/storage/providers/amazon.js]: Fixed

File: server/shared/storage/proxy/providers/cloudfront.js (hunk 6 above)
Comment (jtoric) [reply server/shared/storage/proxy/providers/cloudfront.js]: Fixed

File: server/shared/storage/providers/amazon.js [hunk 7]
```diff
@@ -44,90 +58,120 @@ class Amazon {
...
+  // API docs: https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/classes/getobjectcommand.html
   getFileUrl(key, options = {}) {
     const expires = options.expires || DEFAULT_EXPIRATION_TIME;
-    const params = Object.assign(options, { Bucket: this.bucket, Key: key, Expires: expires });
-    return this.client.getSignedUrlPromise('getObject', params);
+    const command = new GetObjectCommand(
+      Object.assign(options, { Bucket: this.bucket, Key: key })
+    );
+    const request = createRequest(this.client, command);
+    return formatUrl(request, { expiresIn: expires });
```
Comment (droguljic) [inline server/shared/storage/providers/amazon.js]: Couldn't we just use `getSignedUrl`, https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/functions/_aws_sdk_s3_request_presigner.getSignedUrl.html

File: server/shared/storage/providers/amazon.js [hunk 8]
```diff
@@ -44,90 +58,120 @@ class Amazon {
...
+    );
+    return this.client.send(new CopyObjectCommand(params));
   }
 
   moveFile(key, newKey, options = {}) {
-    return this.copyFile(key, newKey, options)
-      .then(result => this.deleteFile(key).then(() => result));
+    return this.copyFile(key, newKey, options).then(result =>
+      this.deleteFile(key).then(() => result)
+    );
```
Comment (droguljic) [inline server/shared/storage/providers/amazon.js]: I don't see any change here, just different code formatting. Remove this and all other code formatting updates, as they obscure what was actually changed.

File: server/shared/storage/providers/amazon.js [hunk 9]
```diff
@@ -44,90 +58,120 @@ class Amazon {
...
+    const upload = new Upload({
+      client: this.client,
+      params: Object.assign(options, {
+        Bucket: this.bucket,
+        Key: key,
+        Body: throughStream
+      })
+    });
+
+    upload.done().catch(noop);
```
Comment (droguljic) [inline server/shared/storage/providers/amazon.js]: Leave formatting as it was, just update with new logic.
```
const params = ...
const upload = new Upload({ client: this.client, params });
upload.done().catch(noop);
```

File: server/shared/storage/providers/amazon.js [hunk 10]
```diff
@@ -22,17 +34,19 @@ class Amazon {
...
     const s3Config = {
-      signatureVersion: 'v4',
-      accessKeyId: config.key,
-      secretAccessKey: config.secret,
+      credentials: {
+        accessKeyId: config.key,
+        secretAccessKey: config.secret
+      },
       region: config.region,
+      signatureVersion: 'v4',
```
Comment (droguljic) [inline server/shared/storage/providers/amazon.js]: Move to top of an object, to reduce diff.

**Review by droguljic — CHANGES_REQUESTED** [review]
Revert all formatting changes, in order for the diff, represents actually implementation change.

File: server/shared/storage/providers/amazon.js (hunk 7 above)
Comment (jtoric) [reply server/shared/storage/providers/amazon.js]: Replaced with getSignedUrl

File: server/shared/storage/providers/amazon.js (hunk 8 above)
Comment (jtoric) [reply server/shared/storage/providers/amazon.js]: Reverted code formatting

File: server/shared/storage/providers/amazon.js (hunk 9 above)
Comment (jtoric) [reply server/shared/storage/providers/amazon.js]: Reverted code formatting

File: server/shared/storage/providers/amazon.js (hunk 10 above)
Comment (jtoric) [reply server/shared/storage/providers/amazon.js]: Reverted code formatting

**Review by droguljic — APPROVED** [review]

**Comment by bubafinder** [comment]
It was tested with a local proxy. Everything works as expected!

**Review by bubafinder — APPROVED** [review]

**Pull request merged**